From df73d16041b01ac55b4aa0932e6b63a719f30353 Mon Sep 17 00:00:00 2001
From: zhengq <qi.zheng@intel.com>
Date: Fri, 10 Oct 2014 17:44:13 +0800
Subject: [PATCH 166/639] I2C ACPI: fix null pointer bug.

Sometimes the value of "length" got in the dw_i2c_acpi_space_handler
is smaller than 2 which may cause the system crash in the following call into
i2c_transfer or memcpy.
Add check for the "length" to prevent this issue.

Change-Id: Ib28024fb7fd06fcd30a35753f07fcba90ea62582
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-2430
Signed-off-by: zhengq <qi.zheng@intel.com>
(cherry picked from commit 2ca6d548229dd4003506a1319b34230dfd3cb47e)

Signed-off-by: Brett T. Warden <brett.t.warden@intel.com>
---
 drivers/i2c/busses/i2c-designware-platdrv.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/i2c/busses/i2c-designware-platdrv.c b/drivers/i2c/busses/i2c-designware-platdrv.c
index ea6f3a1..511b1a9 100644
--- a/drivers/i2c/busses/i2c-designware-platdrv.c
+++ b/drivers/i2c/busses/i2c-designware-platdrv.c
@@ -48,6 +48,8 @@
 
 #include <asm/intel_mid_pcihelpers.h>
 
+#define I2C_TRAN_HDR_SIZE	2
+
 static struct i2c_algorithm i2c_dw_algo = {
 	.master_xfer	= i2c_dw_xfer,
 	.functionality	= i2c_dw_func,
@@ -132,6 +134,9 @@ dw_i2c_acpi_space_handler(u32 function, acpi_physical_address address,
 	if (!value64)
 		return AE_BAD_PARAMETER;
 
+	if (length <= I2C_TRAN_HDR_SIZE)
+		return AE_OK;
+
 	function &= ACPI_IO_MASK; 
 	if (function == ACPI_READ) {
 		buffer = kzalloc(length, GFP_KERNEL);
@@ -152,7 +157,7 @@ dw_i2c_acpi_space_handler(u32 function, acpi_physical_address address,
 			return AE_ERROR;		
 		}
 	
-		memcpy(value + 2, buffer, length - 2);
+		memcpy(value + 2, buffer, length - I2C_TRAN_HDR_SIZE);
 		value[0] = value[1] = 0;
 		kfree(buffer);
 	} else if (function == ACPI_WRITE) {
-- 
2.10.1

